### PR TITLE
Vst2 fix metering

### PR DESCRIPTION
**Pull request by pietro68**
Resolves: metering for vst2

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [ ] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

**Comment by Paul-Licameli** [comment]
Before anything, please rebase this onto latest release-3.2.2 so the number of commits isn't 74!

File: src/effects/VST/VSTEffect.cpp [hunk 1]
```diff
@@ -3790,9 +3790,15 @@ void VSTEffectWrapper::SizeWindow(int w, int h)
 
 void VSTEffectWrapper::Automate(int index, float value)
 {
+   // This is now called only for the instance;
+   // TODO (?) we'd need to copy here part of what is done in VSTEffectValidator::Automate
+   // - we need to send back to the framework the new value for the parameter
```
Comment (Paul-Licameli) [inline src/effects/VST/VSTEffect.cpp:3791]: The commit that adds this comment is making a promise to fix things in a later commit.  I'm still reviewing to see whether that happens.

File: src/effects/VST/VSTEffect.h [hunk 2]
```diff
@@ -100,7 +100,14 @@ struct VSTEffectSettings
 };
 
 
-struct VSTEffectWrapper : public VSTEffectLink, public XMLTagHandler
+struct VSTEffectUIWrapper
+{
+   virtual void NeedIdle();
+   virtual void SizeWindow(int w, int h);
```
Comment (Paul-Licameli) [inline src/effects/VST/VSTEffect.h:106]: I think this interface is insufficient.  I think `NeedEditIdle()` and `Automate()` must be part of it too.

**Comment by Paul-Licameli** [comment]
Another omission so far -- not specific to metering, but needed for a proper implementation of the validator -- is to rewrite the OnSlider event handler for plain UI.

It should contain a call to `ModifySettings` which makes a message from `i` and the slider value and returns that.

It should not do `callSetParameter` directly on the instance.  Let that happen in `RealtimeProcessStart` on the other thread.

File: src/effects/VST/VSTEffect.cpp (hunk 1 above)
Comment (pietro68) [reply src/effects/VST/VSTEffect.cpp]: ok, done now

File: src/effects/VST/VSTEffect.h (hunk 2 above)
Comment (pietro68) [reply src/effects/VST/VSTEffect.h:106]: now done

**Comment by pietro68** [comment]
> 



> Another omission so far -- not specific to metering, but needed for a proper implementation of the validator -- is to rewrite the OnSlider event handler for plain UI.
> 
> It should contain a call to `ModifySettings` which makes a message from `i` and the slider value and returns that.
> 
> It should not do `callSetParameter` directly on the instance. Let that happen in `RealtimeProcessStart` on the other thread.

Your request makes sense, but I think that to do that I'd need a scratch instance and call callSetParameter on it, to be able to correctly update the plainUI.

Why:
Imagine bringing up the GUI and tweaking the "Enable" slider on DeeGain - either when doing destructive processing or when doing realtime processing at play stopped.

As I do this, I expect to see the enable value switch between "On" and "Off", and those strings come from an mAEffect. They are hardcoded in there, and depend on the value which is set to the mAEffect.

**Comment by Paul-Licameli** [comment]
Superseded by #3886 which contains all the same code but a simplified commit history

**Pull request closed without merging**